service/history/historyEngine.go

```diff
@@ -972,6 +972,18 @@ Update_History_Loop:
 
 		scheduleID := token.ScheduleID
 
+		if scheduleID == common.EmptyEventID { // When called by client CompleteActivityById
+			activityID := token.ActivityID
+			if activityID == "" {
+				return &workflow.BadRequestError{Message: "Neither ActivityID nor ScheduleID is provided"}
+			}
+			var ok bool
+			scheduleID, ok = msBuilder.pendingActivityInfoByActivityID[activityID]
```

Comment:
msBuilder has API to lookup ActivityInfo by scheduleID or activityID.  Use the provided API rather than directly accessing the internal data structure.